core/src/impl/Dia/Kokkos/Morpheus_Multiply_Impl.hpp

```diff
@@ -43,58 +50,80 @@ struct DiaSpmv_Kokkos {
   using array_type2 = typename MatrixOrVector2::value_array_type;
   using index_type  = typename LinearOperator::index_type;
   using value_type  = typename LinearOperator::value_type;
+  using member_type = typename Kokkos::TeamPolicy<execution_space>::member_type;
 
-  value_array_type values;
-  array_type1 x;
-  array_type2 y;
-  index_array_type diagonal_offsets;
-  index_type ndiag, ncols;
-
-  DiaSpmv_Kokkos(LinearOperator A, MatrixOrVector1 _x, MatrixOrVector2 _y)
-      : values(A.values.const_view()),
-        diagonal_offsets(A.diagonal_offsets.const_view()),
-        x(_x.const_view()),
-        y(_y.view()) {
-    ndiag = A.values.ncols();
-    ncols = A.ncols();
-  }
-
-  KOKKOS_INLINE_FUNCTION void operator()(const member_type& team_member) const {
-    const index_type row = team_member.league_rank();
-
-    value_type sum = 0;
-    Kokkos::parallel_reduce(
-        Kokkos::TeamThreadRange(team_member, ndiag),
-        [=](const int& n, value_type& lsum) {
-          const index_type col = row + diagonal_offsets[n];
+  const value_array_type values           = A.values.const_view();
+  const array_type1 x_view                = x.const_view();
+  const index_array_type diagonal_offsets = A.diagonal_offsets.const_view();
+  array_type2 y_view                      = y.view();
+  index_type ndiag = A.values.ncols(), ncols = A.ncols();
 
-          if (col >= 0 && col < ncols) {
-            lsum += values(row, n) * x[col];
-          }
-        },
-        sum);
-    y[row] = sum;
-  }
-};
+  const Kokkos::TeamPolicy<execution_space> policy(A.nrows(), Kokkos::AUTO,
+                                                   Kokkos::AUTO);
+
+  Kokkos::parallel_for(
+      policy, KOKKOS_LAMBDA(const member_type& team_member) {
+        const index_type row = team_member.league_rank();
+
+        value_type sum = 0;
+        Kokkos::parallel_reduce(
+            Kokkos::TeamThreadRange(team_member, ndiag),
+            [=](const int& n, value_type& lsum) {
+              const index_type col = row + diagonal_offsets[n];
+
+              if (col >= 0 && col < ncols) {
+                lsum += values(row, n) * x_view[col];
+              }
+            },
+            sum);
+        y_view[row] = sum;
+      });
+}
```

Comment:
Need to check if this has race conditions i.e if hierarchical parallelism was correctly implemented.